api-authentication/server/index.js

```diff
@@ -0,0 +1,85 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+  const sql = `
+  select "userId", "hashedPassword"
+  from "users"
+  where "username" = $1
+  `;
+  const params = [username];
+  db.query(sql, params)
+    .then(result => {
+      if (!result.rows[0]) {
+        throw new ClientError(401, 'invalid login');
+      }
+      const { userId, hashedPassword } = result.rows[0];
+      argon2
+        .verify(hashedPassword, password)
+        .then(passwordMatch => {
+          if (!passwordMatch) {
+            throw new ClientError(401, 'invalid login');
+          }
+          const payload = {
+            userId,
+            hashedPassword
+          };
```

Comment:
Here the payload should have the `userId` and `username` in it only. We never want to send the `hashedPassword` to the client.